### PR TITLE
✨ Add video interface for amp-brightcove

### LINKED ISSUE #7795: Implement the video-interface for amp-brightcove
The [video-interface](https://github.com/ampproject/amphtml/blob/master/src/video-interface.js) is a common API that AMP expects all video players to implement. This allows AMP to treat all video players the same and to use it to provide common UX across players such as autoplay, dockable videos, managing playback across players, etc... 

The interface is fairly straightforward: common methods and events such as `play`, `pause`, `mute`, etc..

- [ ] `amp-brightcove` to implement the  [video-interface](https://github.com/ampproject/amphtml/blob/master/src/video-interface.js). Please look at [amp-youtube.js](https://github.com/ampproject/amphtml/blob/master/extensions/amp-youtube/0.1/amp-youtube.js) for a reference implementation which uses `postMessage` to do the communication between the AMP component and the iframed video player to implement the interface. [amp-ooyala-player](https://github.com/ampproject/amphtml/blob/master/extensions/amp-ooyala-player/0.1/amp-ooyala-player.js) is another player that also implements the interface.
- [ ] `amp-brightcove` NOT to expose `data-param-autoplay`. AMP will manage autoplay for you (exposed as just `autoplay` attribute) if you implement the video-interface. This will ensure consistent autoplay UX across video players. ([Here is demo](https://amphtml-ae5fc.firebaseapp.com/resize/test/manual/amp-videoplayers-autoplay.html) of how AMP handles autoplay).
- [ ] `data-param-playsinline` should be enforced if `autoplay` attribute is present. (Ideally `data-param-playsinline` is the default to be consistent with other players)
- [ ]  `amp-brightcove` is added it to the video-players integration test suite by adding a few lines to [video-players.html](https://github.com/ampproject/amphtml/blob/master/test/fixtures/video-players.html) and [test-video-players.js](https://github.com/ampproject/amphtml/blob/master/test/integration/test-video-players.js)

/cc @mister-ben

### PR DESCRIPTION
**Pull request by mister-ben**
- Implements video interface
- Closes #7795
- Updates docs with new demo players

### CODE REVIEW

**Comment by aghassemi** [comment]
Thanks @mister-ben.
/to @alanorozco for review

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 1]
```diff
@@ -39,21 +82,128 @@ class AmpBrightcove extends AMP.BaseElement {
...
   /** @override */
   buildCallback() {
     this.iframe_ = null;
+
+    this.playerReadyPromise_ = new Promise(resolve => {
+      this.playerReadyResolver_ = resolve;
+    });
+
+    // Warn if the player does not have video interface support
+    this.readyTimeout_ = window.setTimeout(() => {
```
Comment (alanorozco) [inline extensions/amp-brightcove/0.1/amp-brightcove.js]: Please use `Services.timerFor(win).delay`. Easier to test!

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 2]
```diff
@@ -39,21 +82,128 @@ class AmpBrightcove extends AMP.BaseElement {
...
+    this.element.dispatchCustomEvent(VideoEvents.VISIBILITY, {
+      visible,
+    });
+  }
+
   /** @override */
   buildCallback() {
     this.iframe_ = null;
+
+    this.playerReadyPromise_ = new Promise(resolve => {
```
Comment (alanorozco) [inline extensions/amp-brightcove/0.1/amp-brightcove.js]: Use `Deferred` from `src/utils/promise`

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 3]
```diff
@@ -39,21 +82,128 @@ class AmpBrightcove extends AMP.BaseElement {
...
+      getData(event) :
+      tryParseJson(getData(event)));
+    if (data === undefined) {
+      return; // We only process valid JSON.
+    }
+
+    if (data['event']) {
+      if (data['event'] === 'ready') {
+        // Clear warning timeout
+        window.clearTimeout(this.readyTimeout_);
```
Comment (alanorozco) [inline extensions/amp-brightcove/0.1/amp-brightcove.js]: Ditto.

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 4]
```diff
@@ -39,21 +82,128 @@ class AmpBrightcove extends AMP.BaseElement {
...
+        dev().info(TAG, `Player ${this.playerId_} ready. ` +
+          `Brightcove Player version: ${data['bcVersion']} ` +
+          `AMP Support version: ${data['ampSupportVersion']}`);
+        this.hasAmpSupport_ = true;
+        installVideoManagerForDoc(this.element);
+        Services.videoManagerForDoc(this.element).register(this);
+        this.element.dispatchCustomEvent(VideoEvents.LOAD);
+      }
+
+      if (data['event'] === 'playing') {
```
Comment (alanorozco) [inline extensions/amp-brightcove/0.1/amp-brightcove.js]: `return` inside each event conditional.

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 5]
```diff
@@ -39,21 +82,128 @@ class AmpBrightcove extends AMP.BaseElement {
...
+    this.iframe_.contentWindow. /*OK*/ postMessage(JSON.stringify(dict({
+      'command': command,
+      'args': arg,
+    })), 'https://players.brightcove.net');
+  }
+
+  /** @private */
+
+  handlePlayerMessages_(event) {
+    if (event.origin != 'https://players.brightcove.net' ||
```
Comment (alanorozco) [inline extensions/amp-brightcove/0.1/amp-brightcove.js]: nit: Assign `this.element` to a variable in every place where you use it more than once. Makes the file slightly smaller.

**Comment by aghassemi** [comment]
@alanorozco merge-able for this week's Canary?

**Comment by alanorozco** [comment]
@aghassemi Just waiting for Travis!

**Comment by alanorozco** [comment]
@mister-ben Would you mind merging from master, see if that fixes it?

**Comment by aghassemi** [comment]
@mister-ben Interesting, since this PR is not touching core files, it might be a rounding error. Could you please update `cont maxSize` in `bundle-size.js` to `77.02KB` ? /cc @choumx

**Comment by aghassemi** [comment]
@mister-ben I don't seem to be able to push to your PR (not sure why, this used to work) anyway, could you please rebase?

**Comment by aghassemi** [comment]
Thanks again for implementing this @mister-ben , PR is merged.

**Pull request merged**